### PR TITLE
CORS-RFC1918 feedback article #3988

### PR DESCRIPTION
**Pull request by agektmr**
Initial draft to address #3988 CORS-RFC1918 feedback article.

### CODE REVIEW

File: src/site/content/en/blog/cors-rfc1918-feedback/index.md [hunk 1]
```diff
@@ -0,0 +1,168 @@
+---
+title: \[Feedback wanted\] CORS for private networks (RFC1918)
```
Comment (mihajlija) [inline src/site/content/en/blog/cors-rfc1918-feedback/index.md:2]:
```suggestion
title: Feedback wanted: CORS for private networks (RFC1918)
```

File: src/site/content/en/blog/cors-rfc1918-feedback/index.md [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+Malicious websites making requests to devices and servers hosted on a private
+network have long been a threat. Attackers may, for example, change a wireless
+router's configuration to enable
+[Man-in-the-Middle](https://en.wikipedia.org/wiki/Man-in-the-middle_attack)
+attacks. CORS-RFC1918 is a proposal to block such requests by default on the
+browser and require internal devices to opt-in to requests from the public
+internet.
+
+We are looking for feedback from developers who build servers for a private
+network and see how this change will impact the ecosystem.
```
Comment (mihajlija) [inline src/site/content/en/blog/cors-rfc1918-feedback/index.md:26]:
```suggestion
To understand how this change impacts the web ecosystem, we are looking for feedback from developers who build servers for private networks.
```

File: src/site/content/en/blog/cors-rfc1918-feedback/index.md [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+Affected requests include:
+* Requests from the public network to a private network
+* Requests from a private network to a local network
+* Requests from the public network to a local network
+
+**A private network**  
+The destination that resolves to the private address space defined in Section 3
+of [RFC1918](https://tools.ietf.org/html/rfc1918) in IPv4, an IPv4-mapped IPv6
+address where the mapped IPv4 address is itself private, IPv6 subnets excluding
+`::1/128`, `2000::/3` and `ff00::/8`
```
Comment (mihajlija) [inline src/site/content/en/blog/cors-rfc1918-feedback/index.md:84]:
```suggestion
`::1/128`, `2000::/3` and `ff00::/8`.
```

File: src/site/content/en/blog/cors-rfc1918-feedback/index.md [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+
+## What kinds of requests are affected?
+
+Affected requests include:
+* Requests from the public network to a private network
+* Requests from a private network to a local network
+* Requests from the public network to a local network
+
+**A private network**  
+The destination that resolves to the private address space defined in Section 3
```
Comment (mihajlija) [inline src/site/content/en/blog/cors-rfc1918-feedback/index.md:81]:
```suggestion
A destination that resolves to the private address space defined in Section 3
```

File: src/site/content/en/blog/cors-rfc1918-feedback/index.md (hunk 1 above)
Comment (agektmr) [reply src/site/content/en/blog/cors-rfc1918-feedback/index.md:2]: This causes an issue: https://github.com/GoogleChrome/web.dev/issues/4194

File: src/site/content/en/blog/cors-rfc1918-feedback/index.md [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+of [RFC1918](https://tools.ietf.org/html/rfc1918) in IPv4, an IPv4-mapped IPv6
+address where the mapped IPv4 address is itself private, IPv6 subnets excluding
+`::1/128`, `2000::/3` and `ff00::/8`
+
+**A local network**  
+The destination that resolves to the "loopback" space (`127.0.0.0/8`) defined in
```
Comment (mihajlija) [inline src/site/content/en/blog/cors-rfc1918-feedback/index.md:87]:
```suggestion
A destination that resolves to the "loopback" space (`127.0.0.0/8`) defined in
```

File: src/site/content/en/blog/cors-rfc1918-feedback/index.md (hunk 1 above)
Comment (agektmr) [reply src/site/content/en/blog/cors-rfc1918-feedback/index.md:2]: Removing `[` and `[` sounds fair enough

File: src/site/content/en/blog/cors-rfc1918-feedback/index.md [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+prefix (`fe80::/10`) defined in section 2.5.6 of
+[RFC4291](https://tools.ietf.org/html/rfc4291) of IPv6.
+
+**A public network**  
+All others.
+
+<figure class="w-figure">
+  <img src="diagram.png" alt="Relationship between public, private, local networks in CORS-RFC1918">
+  <figcaption class="w-figcaption w-figcaption--fullbleed">
+    Relationship between public, private, local networks in CORS-RFC1918
```
Comment (mihajlija) [inline src/site/content/en/blog/cors-rfc1918-feedback/index.md:102]:
```suggestion
    Relationship between public, private, local networks in CORS-RFC1918.
```

File: src/site/content/en/blog/cors-rfc1918-feedback/index.md [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+  </figcaption>
+</figure>
+
+
+## Chrome's plans to enable CORS-RFC1918
+
+Chrome is bringing CORS-RFC1918 in two steps:
+
+### 1. Requests to private network resources will be allowed only from HTTPS web pages
```
Comment (mihajlija) [inline src/site/content/en/blog/cors-rfc1918-feedback/index.md:111]:
```suggestion
### Step 1: Requests to private network resources will be allowed only from HTTPS web pages
```

File: src/site/content/en/blog/cors-rfc1918-feedback/index.md [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+
+Chrome 87 adds a flag that mandates public websites making requests to private
+network resources to be on HTTPS. You can go to
+`chrome://flags#block-insecure-private-network-requests` to enable it. With this
+flag turned on, any requests to a private network resource from an HTTP website
+will be blocked. The failed requests will be surfaced in the DevTools console
+with the following error code `net::ERR_INSECURE_PRIVATE_NETWORK_REQUEST`.
+
+<figure class="w-figure">
+  <img class="w-screenshot w-screenshot-filled" src="console-error.png" alt="The failed requests will be surfaced in the DevTools console with the error code: `net::ERR_INSECURE_PRIVATE_NETWORK_REQUEST`.">
```
Comment (mihajlija) [inline src/site/content/en/blog/cors-rfc1918-feedback/index.md:121]:
```suggestion
  <img class="w-screenshot w-screenshot-filled" src="console-error.png" alt="Failed requests showing up in the DevTools console with the error code: `net::ERR_INSECURE_PRIVATE_NETWORK_REQUEST`.">
```

File: src/site/content/en/blog/cors-rfc1918-feedback/index.md [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+
+Chrome is bringing CORS-RFC1918 in two steps:
+
+### 1. Requests to private network resources will be allowed only from HTTPS web pages
+
+Chrome 87 adds a flag that mandates public websites making requests to private
+network resources to be on HTTPS. You can go to
+`chrome://flags#block-insecure-private-network-requests` to enable it. With this
+flag turned on, any requests to a private network resource from an HTTP website
+will be blocked. The failed requests will be surfaced in the DevTools console
```
Comment (mihajlija) [inline src/site/content/en/blog/cors-rfc1918-feedback/index.md:117]:
```suggestion
will be blocked. Failed requests will show up in the DevTools console
```

File: src/site/content/en/blog/cors-rfc1918-feedback/index.md [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+with the following error code `net::ERR_INSECURE_PRIVATE_NETWORK_REQUEST`.
+
+<figure class="w-figure">
+  <img class="w-screenshot w-screenshot-filled" src="console-error.png" alt="The failed requests will be surfaced in the DevTools console with the error code: `net::ERR_INSECURE_PRIVATE_NETWORK_REQUEST`.">
+  <figcaption class="w-figcaption w-figcaption--fullbleed">
+    The failed requests will be surfaced in the DevTools console with the error code: `net::ERR_INSECURE_PRIVATE_NETWORK_REQUEST`.
```
Comment (mihajlija) [inline src/site/content/en/blog/cors-rfc1918-feedback/index.md:123]:
```suggestion
    Failed requests showing up in the DevTools console with the error code: `net::ERR_INSECURE_PRIVATE_NETWORK_REQUEST`.
```

File: src/site/content/en/blog/cors-rfc1918-feedback/index.md [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+  </figcaption>
+</figure>
+
+### 2. Sending preflight requests with a special header
```
Comment (mihajlija) [inline src/site/content/en/blog/cors-rfc1918-feedback/index.md:127]:
```suggestion
### Step 2: Sending preflight requests with a special header
```

File: src/site/content/en/blog/cors-rfc1918-feedback/index.md [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+actual request.
+
+The request will include an `Access-Control-Request-Private-Network: true`
+header in addition to other CORS request headers. Among other things, these
+headers identify the origin making the request, allowing for fine-grained access
+control. The server can respond with an `Access-Control-Allow-Private-Network:
+true` header to explicitly indicate that it grants access to the resource.
+
+{% Aside %}
+These headers are still in flux and may change in the future. No action is
```
Comment (mihajlija) [inline src/site/content/en/blog/cors-rfc1918-feedback/index.md:140]:
```suggestion
These headers are still under development and may change in the future. No action is
```

File: src/site/content/en/blog/cors-rfc1918-feedback/index.md [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+
+## Feedback wanted
+
+If you are hosting a website within a private network that expects requests from
+public networks, Chrome team is interested in your feedback and use cases. There
+are two things you can do to help:
+
+* Go to `chrome://flags#block-insecure-private-network-requests` and turn on the
+  flag and see if your website sends requests to the private network resource as
+  expected.
```
Comment (mihajlija) [inline src/site/content/en/blog/cors-rfc1918-feedback/index.md:152]:
```suggestion
* Go to `chrome://flags#block-insecure-private-network-requests`, turn on the
  flag and see if your website sends requests to the private network resource as
  expected.
```

File: src/site/content/en/blog/cors-rfc1918-feedback/index.md [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+* If you encounter any issues or you have any feedback, please file an issue at
+  [crbug.com](https://bugs.chromium.org/p/chromium/issues/entry?components=Blink%3ESecurityFeature%3ECORS)
+  with a component `Blink>SecurityFeature>CORS>RFC1918`.
```
Comment (mihajlija) [inline src/site/content/en/blog/cors-rfc1918-feedback/index.md:155]:
```suggestion
* If you encounter any issues or have feedback, please file an issue at
  [crbug.com](https://bugs.chromium.org/p/chromium/issues/entry?components=Blink%3ESecurityFeature%3ECORS)
  with a component `Blink>SecurityFeature>CORS>RFC1918`.
```

File: src/site/content/en/blog/cors-rfc1918-feedback/index.md [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+> Our wireless router serves an admin website for the same private network but
+> through HTTP. If HTTPS is required for websites that embed the admin website,
+> it will be mixed content. Should we enable HTTPS on the admin website in a
+> closed network?
+
+This is exactly the type of feedback we are looking for. Please file an issue
+with your concrete use case at crbug.com. We'd love to hear from you.
+
+[Hero image](https://unsplash.com/photos/tN344soypQM) by [Stephen
+Philips](https://unsplash.com/@hostreviews) on [Unsplash](https://unsplash.com)
```
Comment (mihajlija) [inline src/site/content/en/blog/cors-rfc1918-feedback/index.md:168]:
```suggestion
[Hero image](https://unsplash.com/photos/tN344soypQM) by [Stephen
Philips](https://unsplash.com/@hostreviews) on [Unsplash](https://unsplash.com).
```

File: src/site/content/en/blog/cors-rfc1918-feedback/index.md [hunk 16]
```diff
@@ -0,0 +1,167 @@
...
+will need to explicitly state that it accepts the cross-origin request by
+responding with corresponding CORS headers. (The exact [CORS
+headers](https://wicg.github.io/cors-rfc1918/) are still under development.)
+
+Developers of such devices or servers will be requested to do two things:
+
+* Make sure the website making requests to a private network is served over
+  HTTPS.
+* Set up the server support for CORS-RFC1918 and respond with expected HTTP
+  headers.
```
Comment (mihajlija) [inline src/site/content/en/blog/cors-rfc1918-feedback/index.md:70]:
```suggestion
To mitigate the threat of similar attacks, the web community is bringing
[CORS-RFC1918](https://wicg.github.io/cors-rfc1918/)—[Cross Origin Resource
Sharing (CORS)](https://developer.mozilla.org/docs/Web/HTTP/CORS) specialized
for private networks defined in [RFC1918](https://tools.ietf.org/html/rfc1918).

Browsers that implement CORS check with target
resources whether they are okay being loaded from a different origin. This is
accomplished either with extra headers inline describing the access or by using
a mechanism called preflight requests, depending on the complexity. Read [Cross
Origin Resource Sharing](https://web.dev/cross-origin-resource-sharing/)
to learn more.

With [CORS-RFC1918](https://wicg.github.io/cors-rfc1918/) the browser will block
loading resources over the private network by default except ones that are
explicitly allowed by the server using CORS and through HTTPS. The website
making requests to those resources will need to send CORS headers and the server
will need to explicitly state that it accepts the cross-origin request by
responding with corresponding CORS headers. (The exact [CORS
headers](https://wicg.github.io/cors-rfc1918/) are still under development.)

Developers of such devices or servers will be requested to do two things:

* Make sure the website making requests to a private network is served over
  HTTPS.
* Set up the server support for CORS-RFC1918 and respond with expected HTTP
  headers.
```

File: src/site/content/en/blog/cors-rfc1918-feedback/index.md [hunk 17]
```diff
@@ -0,0 +1,167 @@
+---
+title: Feedback wanted: CORS for private networks (RFC1918)
+subhead: |
+  Mitigate the risks associated with unintentional exposure of devices and servers on a client’s internal network to the web at large.
+description: |
+  Mitigate the risks associated with unintentional exposure of devices and servers on a client’s internal network to the web at large.
```
Comment (mihajlija) [inline src/site/content/en/blog/cors-rfc1918-feedback/index.md:6]:
```suggestion
  Unintentional exposure of devices and servers on a client’s internal network to the web at large makes them 
  vulnerable to malicious attacks. CORS-RFC1918 is a proposal to block requests from public networks by default 
  on the browser and require internal devices to opt-in to such requests.
```

File: src/site/content/en/blog/cors-rfc1918-feedback/index.md [hunk 18]
```diff
@@ -0,0 +1,181 @@
...
+  <figcaption class="w-figcaption w-figcaption--fullbleed">
+    CORS-RFC1918 errors will be reported as CORS policy errors in the console.
+  </figcaption>
+</figure>
+
+And in the network tab. Note you can filter by **Blocked Requests** to have a
+spotlighted view
+
+<figure class="w-figure">
+  <img class="w-screenshot w-screenshot-filled" src="cors-error.png" alt="CORS-RFC1918 errors will also be reported as CORS error errors in the Network pane.">
```
Comment (mihajlija) [inline src/site/content/en/blog/cors-rfc1918-feedback/index.md:134]:
```suggestion
  <img class="w-screenshot w-screenshot-filled" src="cors-error.png" alt="CORS-RFC1918 errors will also be reported as CORS error errors in the Network panel.">
```

File: src/site/content/en/blog/cors-rfc1918-feedback/index.md [hunk 19]
```diff
@@ -0,0 +1,181 @@
...
+  <figcaption class="w-figcaption w-figcaption--fullbleed">
+    CORS-RFC1918 errors will also be reported as CORS error errors in the Network pane.
```
Comment (mihajlija) [inline src/site/content/en/blog/cors-rfc1918-feedback/index.md:136]:
```suggestion
    CORS-RFC1918 errors will also be reported as CORS error errors in the Network panel.
```

File: src/site/content/en/blog/cors-rfc1918-feedback/index.md [hunk 20]
```diff
@@ -0,0 +1,181 @@
...
+<figure class="w-figure">
+  <img class="w-screenshot w-screenshot-filled" src="console-error.png" alt="CORS-RFC1918 errors will be reported as CORS policy errors in the console.">
+  <figcaption class="w-figcaption w-figcaption--fullbleed">
+    CORS-RFC1918 errors will be reported as CORS policy errors in the console.
+  </figcaption>
+</figure>
+
+And in the network tab. Note you can filter by **Blocked Requests** to have a
+spotlighted view
+
```
Comment (mihajlija) [inline src/site/content/en/blog/cors-rfc1918-feedback/index.md:132]:
```suggestion
And in the **Network** panel. You can filter by **Blocked Requests** to have a
spotlighted view:

```

File: src/site/content/en/blog/cors-rfc1918-feedback/index.md [hunk 21]
```diff
@@ -0,0 +1,181 @@
...
+  Mitigate the risks associated with unintentional exposure of devices and servers on a client’s internal network to the web at large.
+description: |
+  Unintentional exposure of devices and servers on a client’s internal network to the web at large makes them 
+  vulnerable to malicious attacks. CORS-RFC1918 is a proposal to block requests from public networks by default 
+  on the browser and require internal devices to opt-in to such requests.
+authors:
+  - agektmr
+date: 2020-11-09
+hero: hero.jpg
+alt: The hands of a person playing the Chrome dino game on a game console.
```
Comment (letitz) [inline src/site/content/en/blog/cors-rfc1918-feedback/index.md:13]: When I preview the page I see a close-up pic of ethernet ports on what seems to be a router. This seems to be confirmed by the hero image credit at the bottom of the page.

File: src/site/content/en/blog/cors-rfc1918-feedback/index.md [hunk 22]
```diff
@@ -0,0 +1,181 @@
...
+Affected requests include:
+* Requests from the public network to a private network
+* Requests from a private network to a local network
+* Requests from the public network to a local network
+
+**A private network**  
+A destination that resolves to the private address space defined in Section 3
+of [RFC1918](https://tools.ietf.org/html/rfc1918) in IPv4, an IPv4-mapped IPv6
+address where the mapped IPv4 address is itself private, IPv6 subnets excluding
+`::1/128`, `2000::/3` and `ff00::/8`.
```
Comment (letitz) [inline src/site/content/en/blog/cors-rfc1918-feedback/index.md:85]:
```suggestion
of [RFC1918](https://tools.ietf.org/html/rfc1918) in IPv4, an IPv4-mapped IPv6
address where the mapped IPv4 address is itself private, or an IPv6 address outside the
`::1/128`, `2000::/3` and `ff00::/8` subnets.
```

File: src/site/content/en/blog/cors-rfc1918-feedback/index.md [hunk 23]
```diff
@@ -0,0 +1,181 @@
...
+And in the **Network** panel. You can filter by **Blocked Requests** to have a
+spotlighted view:
+
+<figure class="w-figure">
+  <img class="w-screenshot w-screenshot-filled" src="cors-error.png" alt="CORS-RFC1918 errors will also be reported as CORS error errors in the Network panel.">
+  <figcaption class="w-figcaption w-figcaption--fullbleed">
+    CORS-RFC1918 errors will also be reported as CORS error errors in the Network panel.
+  </figcaption>
+</figure>
+
```
Comment (letitz) [inline src/site/content/en/blog/cors-rfc1918-feedback/index.md:139]:
```suggestion

In Chrome 87, CORS-RFC1918 errors are only reported in the DevTools Console as
`ERR_INSECURE_PRIVATE_NETWORK_REQUEST` instead.

You can try it out yourself using this [test website](http://cors-rfc1918-testbed.glitch.me).
```

**Review by letitz — COMMENTED** [review]
Looking great!

File: src/site/content/en/blog/cors-rfc1918-feedback/index.md (hunk 21 above)
Comment (agektmr) [reply src/site/content/en/blog/cors-rfc1918-feedback/index.md:13]: Good catch! Fixed.

File: src/site/content/en/blog/cors-rfc1918-feedback/index.md (hunk 1 above)
Comment (agektmr) [reply src/site/content/en/blog/cors-rfc1918-feedback/index.md]: Looks like `:` is causing Netlify to build at this time. Changing the title...

**Review by mihajlija — APPROVED** [review]

**Comment by agektmr** [comment]
@letitz with your approval we can merge.

**Comment by agektmr** [comment]
@letitz Looks like it's rather a code owner ( @kaycebasques or @jpmedley ) approval that is needed.

File: src/site/content/en/blog/cors-rfc1918-feedback/index.md [hunk 24]
```diff
@@ -10,7 +10,7 @@ authors:
   - agektmr
 date: 2020-11-09
 hero: hero.jpg
-alt: The hands of a person playing the Chrome dino game on a game console.
+alt: A router with eithernet cables connected.
```
Comment (letitz) [inline src/site/content/en/blog/cors-rfc1918-feedback/index.md:13]:
```suggestion
alt: A router with ethernet cables connected.
```

**Review by letitz — APPROVED** [review]
LGTM!

**Comment by letitz** [comment]
I'm surprised the CLA mechanism picked up my github email address (personal) given that I authored no commits. I just changed my primary email address to my @google.com and re-scanned the PR, but that did not change anything. AFAICT there are no commits I authored that I could rebase to fix the situation...

**Comment by agektmr** [comment]
@letitz is a googler and verified he's already signed cla manually.

File: src/site/content/en/blog/cors-rfc1918-feedback/index.md [hunk 25]
```diff
@@ -0,0 +1,189 @@
...
+  on the browser and require internal devices to opt-in to such requests.
+authors:
+  - agektmr
+date: 2020-11-09
+hero: hero.jpg
+alt: A router with ethernet cables connected.
+tags:
+  - blog
+  - security
+---
```
Comment (kaycebasques) [inline src/site/content/en/blog/cors-rfc1918-feedback/index.md:18]:
```suggestion
  - blog
  - security
  - cors
---
```

**Review by kaycebasques — APPROVED** [review]

**Pull request merged**